pkg/frontend/fixetcd.go

```diff
@@ -0,0 +1,669 @@
+	return kubeActions.KubeDelete(ctx, "Job", namespaceEtcds, jobNameDataBackup, true, &propPolicy)
+}
+
+func createBackupEtcdDataJob(cluster, node string) *unstructured.Unstructured {
+	j := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"objectMeta": map[string]interface{}{
+				"name":      jobNameDataBackup,
+				"kind":      "Job",
+				"namespace": namespaceEtcds,
+				"labels":    map[string]string{"app": jobNameDataBackup},
+			},
+			"spec": map[string]interface{}{
+				"template": map[string]interface{}{
+					"objectMeta": map[string]interface{}{
+						"name":      jobNameDataBackup,
+						"namespace": namespaceEtcds,
+						"labels":    map[string]string{"app": jobNameDataBackup},
+					},
+					"activeDeadlineSeconds":   to.Int64Ptr(10),
+					"completions":             to.Int32Ptr(1),
+					"ttlSecondsAfterFinished": to.Int32Ptr(300),
+					"spec": map[string]interface{}{
+						"restartPolicy": corev1.RestartPolicyOnFailure,
+						"nodeName":      node,
+						"containers": []corev1.Container{
+							{
+								Name:  "backup",
+								Image: image,
+								Command: []string{
+									"chroot",
+									"/host",
+									"/bin/bash",
+									"-c",
+									backupOrFixEtcd,
+								},
+								VolumeMounts: []corev1.VolumeMount{
+									{
+										Name:      "host",
+										MountPath: "/host",
+										ReadOnly:  false,
+									},
+								},
+								SecurityContext: &corev1.SecurityContext{
+									Capabilities: &corev1.Capabilities{
+										Add: []corev1.Capability{"SYS_CHROOT"},
+									},
+									Privileged: to.BoolPtr(true),
+								},
+								Env: []corev1.EnvVar{
+									{
+										Name:  "BACKUP",
+										Value: "true",
+									},
+								},
+							},
+						},
+						"volumes": []corev1.Volume{
+							{
+								Name: "host",
+								VolumeSource: corev1.VolumeSource{
+									HostPath: &corev1.HostPathVolumeSource{
+										Path: "/",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// This creates an embedded "metadata" map[string]string{} in the unstructured object
+	// For an unknown reason, creating "metadata" directly in the object doesn't work
+	// and the helper functions must be used
+	j.SetKind("Job")
+	j.SetAPIVersion("batch/v1")
+	j.SetName(jobNameDataBackup)
+	j.SetNamespace(namespaceEtcds)
+
+	return j
+}
+
+// adjustTimeout is used to adjust the amount of time when waiting for jobs to complete when in a testing environment
+func adjustTimeout(ctx context.Context, log *logrus.Entry) time.Duration {
+	if ctx.Value(ctxKey) == "TRUE" {
+		log.Info("Timeout adjusted for testing environment")
+		return time.Microsecond
+	}
+
+	return time.Minute
+}
+
+// comparePodEnvToIp compares the etcd container's environment variables to the pod's actual IP address
+func comparePodEnvToIP(log *logrus.Entry, pods *corev1.PodList) (*degradedEtcd, error) {
+	de := &degradedEtcd{}
+	for _, p := range pods.Items {
```

Comment:
I don't think this for loop will appropriately handle the case where more than two etcd pods have mismatched IPs. It looks like the contents of `de` would be overwritten at each iteration where it encounters an IP mismatch, and then at the end of the loop the RP would move forward trying to fix the most recently found mismatch.

I think you'd want to apply the same idea here as you did in `findCrashloopingPods` where if we find more than one IP mismatch we return an error and quit the recovery attempt.

Reply:
@kimorris27 Thanks for the suggestion. I hadn't considered multiple IP address conflicts, but if more than one node changed NICs this could occur. I've updated this section to handle multiple pod conflicts. I found some issues in the unit tests (degraded pod list wasn't realistic enough) as a result of refracting as well.